Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,598 @@
+            private let maximumConcurrentConnections: Int = 8
+
+            /// Opened connections that are available
+            fileprivate var availableConnections: CircularBuffer<Connection> = .init(initialCapacity: 8)
+
+            /// The number of currently leased connections
+            fileprivate var leased: Int = 0 {
+                didSet {
+                    assert((0...self.maximumConcurrentConnections).contains(self.leased), "Invalid number of leased connections (\(self.leased))")
+                }
+            }
+
+            /// Consumers that weren't able to get a new connection without exceeding
+            /// `maximumConcurrentConnections` get a `Future<Connection>`
+            /// whose associated promise is stored in `Waiter`. The promise is completed
+            /// as soon as possible by the provider, in FIFO order.
+            fileprivate var waiters: CircularBuffer<Waiter> = .init(initialCapacity: 8)
+
+            fileprivate var activity: Activity = .opened
+
+            fileprivate var pending: Int = 0
+
+            private let parentPool: ConnectionPool
+
+            private let key: Key
+
+            fileprivate init(eventLoop: EventLoop, parentPool: ConnectionPool, key: Key) {
+                self.defaultEventLoop = eventLoop
+                self.parentPool = parentPool
+                self.key = key
+            }
+
+            fileprivate mutating func connectionAction(for preference: HTTPClient.EventLoopPreference) -> ConnectionGetAction {
+                self.pending -= 1
+                if self.leased < self.maximumConcurrentConnections {
+                    self.leased += 1
+                    let (channelEL, requiresSpecifiedEL) = self.resolvePreference(preference)
+
+                    if let connection = availableConnections.swapWithFirstAndRemove(where: { $0.channel.eventLoop === channelEL }) {
+                        connection.isLeased = true
+                        return .leaseConnection(connection)
+                    } else {
+                        if requiresSpecifiedEL {
+                            return .makeConnection(channelEL)
+                        } else if let existingConnection = availableConnections.popFirst() {
+                            return .leaseConnection(existingConnection)
+                        } else {
+                            return .makeConnection(self.defaultEventLoop)
+                        }
+                    }
+                } else {
+                    let promise = self.defaultEventLoop.makePromise(of: Connection.self)
+                    self.waiters.append(Waiter(promise: promise, preference: preference))
+                    return .leaseFutureConnection(promise.futureResult)
+                }
+            }
+
+            fileprivate mutating func releaseAction(for connection: Connection) -> ConnectionReleaseAction {
+                if let firstWaiter = waiters.popFirst() {
```

Comment:
missing `self.`

Reply:
@weissi Thank you! I think the NIO projects have this as a requirement in their swift format config file, maybe it could also be added for `AsyncHTTPClient`?